flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/source/eventtimeextractor/IcebergEventTimeExtractor.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.source.eventtimeextractor;
+
+import java.io.Serializable;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+
+/** The interface used to extract watermarks and event timestamps from splits and records. */
+public interface IcebergEventTimeExtractor<T> extends Serializable {
```

Comment:
Flink `TimestampAssigner` also has a nice Javadoc defines the behavior where timestamp is not available.

Reply:
I think the main question is whether we want to have a single interface for watermark extraction and timestamp extraction, or we want 2 interfaces.

If we want to separate the 2 methods, then it might worth to reuse the Flink `TimestampAssigner`. It would still be a bit awkward, as a previous assigned timestamp for the element never exists (so the parameter would be always be `NO_TIMESTAMP), but I see the value of reusing the assigner if possible.